src/hooks/useStyleSheet.ts

```diff
@@ -1,30 +1,36 @@
+/* eslint-disable no-extra-boolean-cast */
 import { RefObject } from "react";
-
 import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect";
+import { getShadowRoot } from "../utils/shadowRoot";
 import { getNonce } from "../utils/nonce";
 
 // Bundler is configured to load this as a processed minified CSS-string
 import styles from "../css/styles.css";
 
-const styleElementMap: Map<Document, HTMLStyleElement> = new Map();
+const styleElementMap: Map<Document | ShadowRoot, HTMLStyleElement> = new Map();
 
 /**
  * Injects CSS code into the document's <head>
  */
 export const useStyleSheet = (nodeRef: RefObject<HTMLDivElement>): void => {
   useIsomorphicLayoutEffect(() => {
-    const parentDocument = nodeRef.current ? nodeRef.current.ownerDocument : document;
+    const parentDocument = nodeRef.current
+      ? !!getShadowRoot(nodeRef.current)
+        ? getShadowRoot(nodeRef.current)
```

Comment:
`getShadowRoot` runs while on nodes tree, so it's kinda expensive operation, so I would suggest running it just once (not 3 times) and putting result into a variable.

```js
const shadowRoot = getShadowRoot(nodeRef.current)
```